src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.util.concurrent.{ExecutorService, Executors}
+
+import scala.collection.mutable
+
+/**
+ * Pool of thread executors. There should be a 1-1 correspondence between C# threads
+ * and Java threads.
+ */
+object ThreadPool {
+
+  /**
+   * Map from threadId to corresponding executor.
+   */
+  val executors: mutable.HashMap[Int, ExecutorService] =
+    new mutable.HashMap[Int, ExecutorService]()
+
+  /**
+   * Run some code on a particular thread.
+   *
+   * @param threadId Integer id of the thread.
+   * @param task Function to run on the thread.
+   */
+  def run(threadId: Int, task: () => Unit): Unit = {
+    val executor = getOrCreateExecutor(threadId)
+    val future = executor.submit(new Runnable {
+      override def run(): Unit = task()
+    })
+
+    future.get()
```

Comment:
I'm curious how `channelRead0` works.   Is each msg started in a new thread?  If so then I think this is okay.  But if channelRead0 reuses a thread that's currently blocking here then I wonder if we would ever hit a deadlock.